From baf2460ff52f722a9d49a239e40c4b53b22c1a4b Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Mon, 21 Jun 2021 09:54:36 +0200
Subject: [PATCH 0637/1695] plat-stm32mp1: add hse monitoring feature

-setup hse monitoring while boot_final level
-register stm32_hse_over_frequency handler on lptimer alarm

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f690a08a4dfd869b3f5a7ecf336c870affec943d plat-stm32mp1: add hse monitoring feature]
Change-Id: Iff3188823c839f1511ce8bc790a83fd05488ff68
Signed-off-by: Ludovic Barre <ludovic.barre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/200760
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Lionel DEBIEVE <lionel.debieve@foss.st.com>
Tested-by: Lionel DEBIEVE <lionel.debieve@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/main.c | 63 ++++++++++++++++++++++++++++++
 1 file changed, 63 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index a4de4bd4a..07ecc5d0c 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -7,6 +7,7 @@
 #include <boot_api.h>
 #include <config.h>
 #include <console.h>
+#include <drivers/counter.h>
 #include <drivers/gic.h>
 #include <drivers/pinctrl.h>
 #include <drivers/stm32_bsec.h>
@@ -653,3 +654,65 @@ void __noreturn do_reset(const char *str __maybe_unused)
 	udelay(100);
 	panic();
 }
+
+#ifdef CFG_STM32_HSE_MONITORING
+/* pourcent rate of hse alarm */
+#define HSE_ALARM_PERCENT	110
+#define FREQ_MONITOR_COMPAT	"st,freq-monitor"
+
+static void stm32_hse_over_frequency(uint32_t ticks __unused,
+				     void *user_data __unused)
+{
+	EMSG("HSE over frequency: nb ticks:%"PRIu32, ticks);
+}
+DECLARE_KEEP_PAGER(stm32_hse_over_frequency);
+
+static TEE_Result stm32_hse_monitoring(void)
+{
+	struct clk *hse_clk = stm32mp_rcc_clock_id_to_clk(CK_HSE);
+	struct clk *hsi_clk = stm32mp_rcc_clock_id_to_clk(CK_HSI);
+	unsigned long hse = 0;
+	unsigned long hsi_cal = 0;
+	struct counter_device *counter;
+	uint32_t ticks = 0;
+	void *fdt = NULL;
+	void *config = NULL;
+	int node = 0;
+
+	DMSG("HSE monitoring");
+
+	fdt = get_embedded_dt();
+	node = fdt_node_offset_by_compatible(fdt, 0, FREQ_MONITOR_COMPAT);
+	if (node < 0)
+		panic();
+
+	if (fdt_get_status(fdt, node) == DT_STATUS_DISABLED)
+		return TEE_SUCCESS;
+
+	hse = clk_get_rate(hse_clk);
+	hsi_cal = clk_get_rate(hsi_clk);
+	/*
+	 * hsi_cal is based on hsi & DIVISOR
+	 * DIVISOR is fixed, (stm32mp13:1024)
+	 */
+	hsi_cal /= 1024;
+
+	ticks = (hse / 100) * HSE_ALARM_PERCENT;
+	ticks /= hsi_cal;
+
+	DMSG("HSE:%luHz HSI cal:%luHz alarm:%"PRIu32, hse, hsi_cal, ticks);
+
+	counter = fdt_counter_get(fdt, node, &config);
+	assert(counter && config);
+
+	counter->alarm.callback = stm32_hse_over_frequency;
+	counter->alarm.ticks = ticks;
+
+	counter_start(counter, config);
+	counter_set_alarm(counter);
+
+	return TEE_SUCCESS;
+}
+
+driver_init_late(stm32_hse_monitoring);
+#endif /* CFG_STM32_HSE_MONITORING */
-- 
2.47.2

